pkg/splunk/enterprise/workerpool.go

```diff
@@ -0,0 +1,184 @@
+// Workers wait for the jobs on the jobs channel and pick them up as they come in
+func (w *AppInstallWorkerPool) Run(id int, jobs <-chan *enterpriseApi.AppDeploymentInfo) {
+	scopedLog := log.WithName("AppInstallWorkerPool.run()").WithValues("worker id", id)
+	for job := range jobs {
+		scopedLog.Info("Started install of app", "appName", job.AppName)
+
+		object := getRemoteObjectFromS3Response(job.AppName, w.S3Reponse)
+		localFile := getLocalAppFileName(w.DownloadPath, job.AppName, *object.Etag)
+		remoteFile := *object.Key
+
+		// download the app from remote storage
+		err := w.S3ClientMgr.DownloadApp(remoteFile, localFile, *object.Etag)
+		if err != nil {
+			scopedLog.Error(err, "unable to download app", "appName", job.AppName)
+			setAppInstallState(job, enterpriseApi.DownloadError)
+		} else {
+			// download is successfull, update the state
+			setAppInstallState(job, enterpriseApi.DownloadComplete)
+
+			//TODO: gaurav, need to fill in below code
+			// copy the app to the splunk pod
+
+			// untar it, if it is cluster scoped
+
+			// if everything is successful, delete the app from operator pod
+		}
+
+		scopedLog.Info("finished job")
+		w.mu.Lock()
+		w.availableDiskSpace = w.availableDiskSpace + job.Size
+		scopedLog.Info("availableDiskSpace after freeing up space", "availableDiskSpace", w.availableDiskSpace)
+		w.mu.Unlock()
+
+		w.wg.Done()
+	}
+}
+
+// isAppAlreadyDownloaded checks if the app is already downloaded on the operator pod by checking a couple of things -
+// 1. is the app bundle present on the operator pod &&
+// 2. if the size on disk matches the size reported from remote storage
+func (w *AppInstallWorkerPool) isAppAlreadyDownloaded(app *enterpriseApi.AppDeploymentInfo) bool {
+	scopedLog := log.WithName("isAppAlreadyDownloaded").WithValues("app name", app.AppName)
+	object := getRemoteObjectFromS3Response(app.AppName, w.S3Reponse)
+	localAppFileName := getLocalAppFileName(w.DownloadPath, app.AppName, *object.Etag)
+
+	var fileInfo fs.FileInfo
+	var err error
+	// 1. First check if the app is present on operator pod
+	if fileInfo, err = os.Stat(localAppFileName); os.IsNotExist(err) {
+		scopedLog.Info("App not present on operator pod")
+		return false
+	}
+
+	// 2. Now, since the app is present, check the size
+	localSize := fileInfo.Size()
+	remoteSize := *object.Size
+	if localSize != remoteSize {
+		err = fmt.Errorf("size of app on operator pod does not match size on remote storage. localSize: %d, remoteSize:%d", localSize, remoteSize)
+		scopedLog.Error(err, "need to re-download app")
+		return false
+	}
+
+	// If we reached here, this means app was downloaded successfuly to operator pod,
+	// mark this as DownloadComplete
+	setAppInstallState(app, enterpriseApi.DownloadComplete)
+	return true
+}
+
+// Start kicks off the workers to do the job of app installation
+func (w *AppInstallWorkerPool) Start() {
+	// start the workers
+	for i := 1; i <= int(w.NumOfWorkers); i++ {
+		go w.Run(i, w.JobsChan)
+	}
```

Comment:
I think we need to handle this in a more controlled way, otherwise the go routines may end prematurely, even before the jobs queue is filled up? It may not be an issue, as long as the main thread is not doing any blocking calls before adding the jobs to the channel.
Alternatively, maybe moved to the line after 171, if that works?. I think addressing it in a more controlled way(wait for the main thread to give an indication) is preferable, such that, we don't need to worry in case even if dealing with parallel executions involving multiple CPUs.

Reply:
Yeah, so that is the interesting thing about channels in goroutine. The workers won't exit prematurely since they are waiting on jobs in the jobs channel. And unless you close the channel they(goroutines) won't exit. And you only close the channel once all the jobs are sent in the channel, hence I think we are safe here.

Reply:
Ahh, then it makes sense.